Set effective_value when initializing a COutput

ishaanam opened the pull request:
Previously in COutput, effective_value was initialized as the absolute value of the txout and the fee as 0. effective_value along with the fee was calculated outside of the COutput constructor and set after the object had been initialized.
These changes will allow either the fee or the feerate to be passed in a COutput constructor and the fee and effective_value are calculated and set in the constructor. As a result, AvailableCoins also needs to be passed the feerate when utxos are being spent. When balance is calculated or the coins are being listed and feerate is neither available nor required, AvailableCoinsListUnspent is used instead, which runs AvailableCoins while providing the default value for `feerate`. Unit tests for the calculation of effective value have also been added.

ishaanam committed:
Set effective_value when initializing a COutput

Previously in COutput, effective_value was initialized as the absolute
value of the txout, and fee as 0. effective_value along with fee were
calculated outside of the COutput constructor and set after the
object had been initialized. These changes will allow either the fee
or the feerate to be passed in a COutput constructor. If either are
provided, fee and effective_value are calculated and set in the
constructor. As a result, AvailableCoins also needs to be passed the
feerate when utxos are being spent. When balance is calculated or the
coins are being listed and feerate is neither available nor required,
AvailableCoinsListUnspent is used instead, which runs AvailableCoins
while providing the default value for feerate. Unit tests for the
calculation of effective value have also been added.

achow101 commented on src/wallet/spend.h:
```diff
@@ -35,7 +35,9 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* walle
-void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, std::optional<CFeeRate> feerate = std::nullopt, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+
+void AvailableCoinsListUnspent(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0);
```
```suggestion
void AvailableCoinsListUnspent(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
```

This should fix the clang error.

ishaanam replied:
Thanks, it's fixed now

achow101 commented on src/wallet/spend.cpp:
```diff
@@ -191,8 +191,12 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
             int input_bytes = GetTxSpendSize(wallet, wtx, i, (coinControl && coinControl->fAllowWatchOnly));
-
-            vCoins.emplace_back(COutPoint(wtx.GetHash(), i), wtx.tx->vout.at(i), nDepth, input_bytes, spendable, solvable, safeTx, wtx.GetTxTime(), tx_from_me);
+            
```
There is extra whitepace on this line that the linter doesn't like.

achow101 commented on src/bench/coin_selection.cpp:
```diff
@@ -58,7 +58,7 @@ static void CoinSelection(benchmark::Bench& bench)
     std::vector<COutput> coins;
     for (const auto& wtx : wtxs) {
-        coins.emplace_back(COutPoint(wtx->GetHash(), 0), wtx->tx->vout.at(0), /*depth=*/6 * 24, GetTxSpendSize(wallet, *wtx, 0), /*spendable=*/true, /*solvable=*/true, /*safe=*/true, wtx->GetTxTime(), /*from_me=*/true);
+        coins.emplace_back(COutPoint(wtx->GetHash(), 0), wtx->tx->vout.at(0), /*depth=*/6 * 24, GetTxSpendSize(wallet, *wtx, 0), /*spendable=*/true, /*solvable=*/true, /*safe=*/true, wtx->GetTxTime(), /*from_me=*/true, /*fee=*/ 0);
```
s/fee/fees

Here and elsewhere. clang-tidy is complaining about this.

ishaanam replied:
I've changed it everywhere, thanks.

achow101 commented on src/wallet/spend.h:
```diff
@@ -35,7 +35,9 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* walle
  * populate vCoins with vector of available COutputs.
  */
-void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, const CAmount& nMinimumAmount = 1, std::optional<CFeeRate> feerate = std::nullopt, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
```
nit: I think it would be better to put `feerate` after `coinControl` to maintain the grouping of similar kinds of parameters.

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -73,16 +75,49 @@ struct COutput {
+        }
+    }
+
+    COutput(const COutPoint& outpoint, const CTxOut& txout, int depth, int input_bytes, bool spendable, bool solvable, bool safe, int64_t time, bool from_me, const CFeeRate feerate)
```
I think it would be better to have `feerate` be a `std::optional<CFeeRate` with a default of `std::nullopt`. This would allow us to do some simplification in `AvailableCoins` where the `CFeeRate` can just be passed in without checking whether it has a value.

ishaanam replied:
I agree, it looks a lot cleaner now.

murchandamus reviewed (commented):
This looks like it's almost ready to go. I have a couple nits about names, that we might need to get some advice from more experienced contributors on, and a couple more potential test cases that I'm not sure what the outcome would be for.

ishaanam commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -815,5 +813,40 @@ BOOST_AUTO_TEST_CASE(waste_test)
+    COutput output4(COutPoint(tx.GetHash(), nInput), tx.vout.at(nInput), /*depth=*/ 1, input_bytes, /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, /*time=*/ 0, /*from_me=*/ false, fees);
+    BOOST_CHECK_EQUAL(output4.GetEffectiveValue(), expected_ev1);
+
+    // input bytes unknown (input_bytes = -1), pass fees in constructor
+    COutput output5(COutPoint(tx.GetHash(), nInput), tx.vout.at(nInput), /*depth=*/ 1, /*input_bytes=*/ -1, /*spendable=*/ true, /*solvable=*/ true, /*safe=*/ true, /*time=*/ 0, /*from_me=*/ false, /*fees=*/ 0);
+    BOOST_CHECK_EQUAL(output5.GetEffectiveValue(), nValue); // The effective value should be equal to the absolute value if input_bytes is -1
```
I think that in both of these cases the assertion on src/wallet/coinselection.h:98 will just fail. `fee` must be `0` whenever `input_bytes` has a negative value.

Xekyo replied:
Ah good to see that you're stating your assumptions. :)

laanwj commented on src/wallet/coinselection.h:
```diff
@@ -73,16 +75,49 @@ struct COutput {
+        }
+    }
+
+    COutput(const COutPoint& outpoint, const CTxOut& txout, int depth, int input_bytes, bool spendable, bool solvable, bool safe, int64_t time, bool from_me, const CAmount fees)
```
> Perhaps we should use this PR as an opportunity to rename the class members of COutput to use the m_ prefix as suggested by the [Style Guide](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c)? 

Dunno. FWIW `m_` meant is for fully-fledged, self contained classes, not 'dumb structs' which simply keep some data together (like `COutput` seems to be). The rationale for it is that within methods it's easier to distinguish local variables from class variables.

Xekyo replied:
Okay, thanks for the guidance

Xekyo commented on src/wallet/coinselection.cpp:
```diff
@@ -380,8 +374,8 @@ CAmount GetSelectionWaste(const std::set<COutput>& inputs, CAmount change_cost,
-        waste += coin.fee - coin.long_term_fee;
-        selected_effective_value += use_effective_value ? coin.effective_value : coin.txout.nValue;
+        waste += coin.GetFee() - coin.long_term_fee;
+        selected_effective_value += use_effective_value ? coin.GetEffectiveValue(): coin.txout.nValue;
```
style nit:
```suggestion
        selected_effective_value += use_effective_value ? coin.GetEffectiveValue() : coin.txout.nValue;
```

Xekyo commented on src/wallet/coinselection.h:
```diff
@@ -20,6 +20,14 @@ static constexpr CAmount CHANGE_UPPER{1000000};
 /** A UTXO under consideration for use in funding a new transaction. */
 struct COutput {
+private:
+    /** The output's value minus fees required to spend it. Initialized as the output's absolute value. */
```
This comment seems to be outdated now. We should be initializing this with the correct value now.

ishaanam replied:
It is outdated, I forgot to update this comment.

Xekyo commented on src/wallet/coinselection.h:
```diff
@@ -73,16 +75,49 @@ struct COutput {
+          from_me{from_me}
+    {
+        // if input_bytes unknown, the fees should be 0, if input_bytes is known, then the fees should be a positive integer
+        assert((input_bytes < 0 && fees == 0) || (input_bytes > 0 && fees >= 0));
```
Optional:
> if input_bytes is known, then the fees should be a positive integer

I see that some of the tests for this constructor pass `fees = 0`. What would need to happen for `fees` to be zero when `input_bytes` is greater than zero? IIRC, when SFFO is being used, we use the actual value of the UTXO instead of its effective value, so there too, a fee greater than zero should follow if both feerate and `input_bytes` are known. So, I wonder here whethere `fees = 0` only occurs in the context of tests, or whether this occurs perhaps also in the context of listing unspents outside of transaction building. If so, it could perhaps be mentioned in the comment here.

ishaanam replied:
Yes, `fees = 0` only happens in the context of tests so far. I will mention this here.

murchandamus reviewed (commented):
Approach ACK b07500f28cff352239a6ffbf1da9f4d2bc965838

I noticed that the lines in the commit message are pretty long, please break lines in the body of the commit message after 72 characters. It would be great if the comment could get updated, but I don't think any of my remarks are blockers.

Xekyo commented:
ACK f5a7495c546c13c985bd92a030af092b17c67108

glozow commented on src/wallet/spend.cpp:
```diff
@@ -211,13 +211,19 @@ void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const C
     }
 }
 
+// Allows the 'feerate' optional parameter of AvailableCoins to be skipped when utxos are not being spent
```
This comment would fit better above the `AvailableCoinsListUnspent` declaration in wallet/spend.h so someone looking to understand how to use the interface at a glance can see it. Also, consider making it [doxygen-compatible](https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-doxygen-compatible-comments).

ishaanam replied:
done, I also clarified the comment a bit as well

ishaanam commented:
At this point, I've addressed all of the suggestions provided by @achow101 @glozow and @Xekyo (thank you!) except for the `fee_rate` nit which I'm still unsure whether to act upon or not.

Xekyo commented:
> At this point, I've addressed all of the suggestions provided by @achow101 @glozow and @Xekyo (thank you!) except for the `fee_rate` nit which I'm still unsure whether to act upon or not.

Unless more people start asking for the `fee_rate` change, don't worry about that. :smile:

achow101 commented on src/wallet/coinselection.h:
```diff
@@ -73,16 +75,49 @@ struct COutput {
+    }
+
+    COutput(const COutPoint& outpoint, const CTxOut& txout, int depth, int input_bytes, bool spendable, bool solvable, bool safe, int64_t time, bool from_me, const CAmount fees)
+        : outpoint{outpoint},
+          txout{txout},
+          depth{depth},
+          input_bytes{input_bytes},
+          spendable{spendable},
+          solvable{solvable},
+          safe{safe},
+          time{time},
+          from_me{from_me}
```
nit: This could be condensed by using constructor delegation

```suggestion
        : COutput(outpoint, txout, depth, input_bytes, spendable, solvable, safe, time, from_me)
```

ishaanam replied:
I will do this the next time I modify this PR

ishaanam replied:
Done

achow101 reviewed (commented):
ACK 1f1a75f3d2277f05549395f80b5e1c2d932a7d3f

Xekyo commented:
re-ACK 1f1a75f3d2277f05549395f80b5e1c2d932a7d3f

glozow commented:
Silent merge conflict with #24820

glozow commented on src/wallet/spend.h:
```diff
@@ -35,7 +35,13 @@ TxSize CalculateMaximumSignedTxSize(const CTransaction& tx, const CWallet* walle
+void AvailableCoins(const CWallet& wallet, std::vector<COutput>& vCoins, const CCoinControl* coinControl = nullptr, std::optional<CFeeRate> feerate = std::nullopt, const CAmount& nMinimumAmount = 1, const CAmount& nMaximumAmount = MAX_MONEY, const CAmount& nMinimumSumAmount = MAX_MONEY, const uint64_t nMaximumCount = 0) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+
+/**
+ * Runs AvailableCoins but allows the `feerate` optional parameter of
```
nit
```suggestion
 * Wrapper function for AvailableCoins which skips the `feerate` parameter. Use this function
 * to list all available coins (e.g. listunspent RPC) while not intending to fund a transaction.
```

ishaanam replied:
Done, this sounds much better

glozow reviewed (commented):
almost ACK, needs a rebase, 1 nit for your consideration

ishaanam commented:
Implemented changes suggested by @glozow and @achow101 and rebased due to silent merge conflict

achow101 commented on src/wallet/test/coinselector_tests.cpp:
```diff
@@ -353,36 +351,67 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
-        coin_selection_params_bnb.m_long_term_feerate = CFeeRate(3000);
         const auto result11 = SelectCoins(*wallet, coins, 10 * CENT, coin_control, coin_selection_params_bnb);
         BOOST_CHECK(EquivalentResult(expected_result, *result11));
+    }
+    {
+        std::unique_ptr<CWallet> wallet = std::make_unique<CWallet>(m_node.chain.get(), "", m_args, CreateMockWalletDatabase());
+        wallet->LoadWallet();
+        LOCK(wallet->cs_wallet);
+        wallet->SetWalletFlag(WALLET_FLAG_DESCRIPTORS);
+        wallet->SetupDescriptorScriptPubKeyMans();
+
+        std::vector<COutput> coins;
+
```
This setup does not need to be repeated for each of these cases. You can just do `coins.clear()` prior to calling `add_coin` again.

ishaanam replied:
Makes sense, I've changed it

achow101 commented:
re-ACK 6fbb0edac22c63f1b723f731c2601b1d46879a58

w0xlt reviewed (approved):
Code Review ACK https://github.com/bitcoin/bitcoin/pull/25083/commits/6fbb0edac22c63f1b723f731c2601b1d46879a58

furszy reviewed (approved):
Looks good, have been touching this area lately, code review ACK 6fbb0eda.

Xekyo commented:
re-ACK 6fbb0edac22c63f1b723f731c2601b1d46879a58

MarcoFalke commented:
Looks like this crashes

https://bugs.chromium.org/p/oss-fuzz/issues/detail?id=47613

https://github.com/bitcoin/bitcoin/commit/1021e4cc6877f66729e01a6662d5877e718ed276
